gettext-rs/src/macros.rs

```diff
@@ -181,73 +224,29 @@ macro_rules! dngettext {
-        }
-        msgstr
-    }};
     ( $domainname:expr, $category:expr,
-      $msgid:expr, $msgid_plural:expr, $n:expr, $($args:expr),+ $(,)? ) => {{
-        let mut msgstr = $crate::dcngettext($domainname, $msgid, $msgid_plural, $n, $category);
-        while msgstr.contains("{n}") {
-            msgstr = $crate::macros::rt_format(&msgstr, "{n}", $n);
-        }
-        $crate::rt_format!(msgstr, "{}", $($args),+)
-    }};
-}
-
-/// Like [`pgettext`], but allows for formatting.
-///
-/// It calls [`pgettext`] on `msgctxt` and `msgid`, and then replaces each occurrence of `{}` with
-/// the next value out of `args`.
-///
-/// # Panics
-///
-/// If compiled with debug assertions enabled (as in "dev" profile),
-/// will panic if the number of arguments doesn't match the number of format directives.
-///
-/// [`pgettext`]: fn.pgettext.html
-#[macro_export]
-macro_rules! pgettext {
-    ( $msgctxt:expr, $msgid:expr $(,)? ) => {
-        $crate::pgettext($msgctxt, $msgid)
-    };
-    ( $msgctxt:expr, $msgid:expr, $($args:expr),+ $(,)? ) => {{
-        $crate::rt_format!(
-            $crate::pgettext($msgctxt, $msgid),
-            "{}",
-            $($args),+
+      $msgid:expr, $msgid_plural:expr, $n:expr $(, $args:expr)* $(,)? ) => {{
+        $crate::macros::rt_format(
+            $crate::dcngettext($domainname, $msgid, $msgid_plural, $n, $category),
+            vec![$(&$args),*],
+            Option::Some(&$n)
         )
     }};
 }
 
-/// Like [`npgettext`], but allows for formatting.
-///
-/// It calls [`npgettext`] on `msgctxt`, `msgid`, `msgid_plural`, and `n`, and then replaces each
-/// occurrence of `{}` with the next value out of `args`, and `{n}` with `n`.
-///
-/// # Panics
-///
-/// If compiled with debug assertions enabled (as in "dev" profile),
-/// will panic if the number of arguments doesn't match the number of format directives.
-///
-/// [`npgettext`]: fn.npgettext.html
-#[macro_export]
-macro_rules! npgettext {
-    ( $msgctxt:expr, $msgid:expr, $msgid_plural:expr, $n:expr $(,)? ) => {{
-        let mut msgstr = $crate::npgettext($msgctxt, $msgid, $msgid_plural, $n);
-        while msgstr.contains("{n}") {
-            msgstr = $crate::macros::rt_format(&msgstr, "{n}", $n);
-        }
-        msgstr
-    }};
-    ( $msgctxt:expr, $msgid:expr, $msgid_plural:expr, $n:expr, $($args:expr),+ $(,)? ) => {{
-        let mut msgstr = $crate::npgettext($msgctxt, $msgid, $msgid_plural, $n);
-        while msgstr.contains("{n}") {
-            msgstr = $crate::macros::rt_format(&msgstr, "{n}", $n);
-        }
-        $crate::rt_format!(msgstr, "{}", $($args),+)
-    }};
+#[cfg(test)]
+mod tests {
+    use macros::rt_format;
+
+    #[test]
+    fn partial_escaping() {
+        assert_eq!(
+            rt_format(
+                String::from("{{}, {}}, {{n}, {n}}"),
+                vec![&"smth"],
+                Option::Some(&5)
+            ),
+            "{, smth, {n, 5"
```

Comment:
Oops, IDK how I haven't thought about adding that debug assert. About the second point — I think it's better to output the least broken string. Having placeholders in place should give users an understanding that something had to be there but somehow isn't. Removing placeholders can break the meaning of the text, or worse — change it. So, I think having placeholders in place is a better decision.

Reply:
It could be argued that an invalid format string doesn't have meaning, so there's nothing to break :) Also, keeping the brace would be inconsistent with how we handle unsupported specifiers (see #85): I'd like us to consume specifiers, so we should consume any other invalid stuff too. Does that make sense, or are you still not convinced? (This issue is partly a matter of taste, so I'm arguing with you in hopes of finding out which position is more defensible :)

Reply:
I take it I failed to convince you here, @delight-aug?

Reply:
Oh, okay, sorry both hurrying you :) I think the draft status is only useful when you first submit a PR, so don't bother with it now. Once you're done editing and CI passes, you can press "Re-request review" next to my nickname in the top right, and I'll get a ping to look at this again.

Thanks for working on this! :+1:

Reply:
Actually, I think I misunderstood you a bit. I agree that macros _should_ swallow unsupported and (*may* swallow) unused directives. However, I disagree with your proposition on escaping. In [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7469ede9f3d63c504a440e0b2bfd7631) you might notice that `{{}` and `{}}` give the same error message "unmatched `}` in format string". It's because in the first case `{{}`, it sees an escaped opening brace and an unmatched unescaped closing brace. And in the second one (`{}}`), it's a valid format directive and an unmatched unescaped closing brace. And so, logically, the output should be `{` and `arg`. I think `debug_assert` should be enough to notice the not-entirely-escaped format directive (`{{}`/`{}}`), and making this use-case specific feature is a bit unlogical, probably won't help a lot, and isn't worth of effort. However, the final decision is up to you, @Minoru.

Reply:
Your examples do make sense, but I feel uneasy about making such an arbitrary decision.

Instead, I looked for some prior art:
* Python's `format` method throws `ValueError` if the format string contains imbalanced curly braces
* C++ `std::vformat` [throws](https://en.cppreference.com/w/cpp/utility/format/vformat) too
* [runtime-fmt](https://crates.io/crates/runtime-fmt) and [dynfmt](https://crates.io/crates/dynfmt) crates return `Result`

I also got thinking about potential users of this feature, and looked through our "dependents" on crates.io:
* GTK apps mentioned in #18. They use Rust-style formats. They assume the format string is correct (i.e. they don't panic and don't return `Result)
* my own app, Newsboat, where I wrap `strnprintf` C function. I use printf-style formats. I assume the format string is correct
* gstore crate [has its own macro](https://gitlab.com/loers/gstore/-/blob/5e367356a7db2254635241224d238fdb0d09570f/src/l10n.rs#L131). It uses Rust-style formats. Assumes the format string is correct
* [tr crate](https://crates.io/crates/tr). Rust-style formats, assumes correctness
* [serde_yaml crate](https://github.com/cecton/serde-gettext) uses Rust-style formats, returns `Result`

I start to think that all this formatting business should be punted onto a separate crate, and gettext-rs should just deal with safe wrappers around gettext functions. Let me sleep on this for a day, I'll get back to you tomorrow. (You can try to persuade me in the meantime :)

Reply:
I think functions in the gettext-rs crate can fulfill the purpose of giving a safe API, while macros can be wrappers for formatting them. We could separate macros to their own crate or feature. Also, we'd better support different formatting backends depending on the dependency used in the binary application. This way, we can have all of the formatting features without hassle. 

However, it's unclear what to do with return types. From the perspective of the user of macros, with Result, I could call auto-bug-reporting and display a message to users, that something went wrong. Or I could just unwrap it all the time and let my program crash because of some string. Also, it doesn't look good when you want to include a phrase in GtkLabel, but you need to match it first (or unwrap it in place). Maybe a good decision is to let users specify an action to run on a broken string? So they set up a closure near textdomain bindings. And by default just swallow things... Just spamming ideas

Reply:
Yeah, my tentative plan is to remove our macros and open an issue explaining what we tried and what problems we encountered. Interested users would comment with their thoughts, perhaps we'll find a new design, which we will then implement. In the meantime, anyone can create new crates on top of gettext-rs to try out ideas.

I'll be sure to include your ideas in that issue :)